Dtls srtp support

RonEld opened the pull request:
## Description
This PR supersedes PR #361 by @jeannotlapin, with additional mki support, some rework, and rebased on current development branch
Supersedes #1540 

**Open question:**
How will the server decide whether to accept the MKI value from the client, in the `use_srtp` extension?  Should there be a callback function? Should the server user set the MKI in advance and the server send it if it matches? Should the server just accept the MKI from the client and use it as its MKI ( basically making `mbedtls_ssl_dtls_srtp_set_mki_value()` a client only function)?

## Status
**READY**

## Requires Backporting
NO  


## Additional comments
Missing still:
~~tests~~
~~interop tests~~

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [ ] Backported


## Steps to test or reproduce
run `ssl-opt.sh -f SRTP`

jeannotlapin commented:
@RonEld,
here are patches to fix the self-signed certificate usage and the need of server certificateVerify messages.
It shall allow interop with webRTC.
First patch just merge the release 2.11.0 into this pr branch (I did it on the PR1540, it shall still be ok)
The second one applies the fix.

I didn't performed intensive testing but it shall work. Still we need tests (ideally interop ones but it's difficult to automatize) for the PR.
Johan

[0001-Change-the-library-version-to-2.11.0.patch.txt](https://github.com/ARMmbed/mbedtls/files/2148865/0001-Change-the-library-version-to-2.11.0.patch.txt)
[0002-DTLS-SRTP-forces-ssl-verify-to-optional-when-not-set.patch.txt](https://github.com/ARMmbed/mbedtls/files/2148866/0002-DTLS-SRTP-forces-ssl-verify-to-optional-when-not-set.patch.txt)

RonEld commented:
@jeannotlapin Thank you for supplying your patches.
I am puzzled with the set authmode to optional part.
The RFC states that the server must send a certificate in dtls-srtp. Why do you think it's ok to keep the authmode optional in this case?

RonEld commented:
@jeannotlapin I believe that the reason you get failure when the server sends their self signed certificate, is because you haven't set the server's trusted root CA certificate through `mbedtls_ssl_conf_ca_chain()`. When you set the authmode to optional, the certificate verification could still fail, but you ignore it.
This is not the desired behavior.
If it's a different issue, please clarify.

jeannotlapin commented:
@RonEld 
The regular DTLS SRTP handshake when interfacing webRTC can't provide the root CA as trusted chain.
webRTC is generating on the fly a self signed certificate (actually it's a temporary one used few days)
Setting the verify to optional will force the optional certificateRequest packet to be sent by the server to follow the RFC as both sides will provide certificate. A certificate fingerprint is sent on the signaling channel during call establishment allowing to confirm the trust on the peer certificate (being client or server) without using the regular TLS certificate chain mechanism.

So when doing a DTLS handshake to produce SRTP keys, it is requested to use client and server certificate but you do not need to verify them during the handshake as they are supposed to be verified through the certificate fingerprint mechanism.

That's why my patch will only upgrade to optional the verify mode when set to none as you may use the DTLS handshake for SRTP in other context and thus use the regular verify mode.

RonEld commented:
@jeannotlapin Thank you for your explanation.
To ease my concerns, could you please refer me to were this certificate fingerprint is mentioned?
Are you referring to [RFC 5763](https://tools.ietf.org/html/rfc5763)?

I am concerned that we allow a handshake with a peer which certificate is not verified successful.
It might work with the specific dtls-srtp infrastructure you are working with, but can you guarantee it will work for all?

roxlu commented:
I tried to use this PR and after applying the patches it indeed succeeds the "verification" but I do receive this error  "ECP - The signature is not valid". It might be that I'm not setting up my self-signed certificate correctly (?). I'm not sure if I need to configure ECP.

jeannotlapin commented:
@RonEld  yes I was referring to RFC5763.
Thing is that the patch will just turn on the certificate exchange from both parts if the user set the verify to none but it won't downgrade to no verification if the user set the ssl_verify_required.

So if a user set the flag to verify_none it will force the use of client and server certificate anyway as requested by the RFC5764 but setting the verify to optional. In case the verifyy flag was set to required the patch will have no effect so it won't make any damage to a scheme explicitely requesting the verification to take place.

The general usage of DTLS-SRTP uses self-signed certificate (or it means that each end user could provide a valid certificate that could be verified by its correspondant but it is nearly never the case when using VOIP).

@roxlu do you think your error is linked to the patch? Was it ok before the patch and without the commit forcing the use of certificate ?

RonEld commented:
@jeannotlapin Thanks for your explanation

@roxlu When the authmode is optional, thenthe handshake will continue, ignoring the verification error, if there was. Could you clarify what API returns this "ECP - The signature is not valid` error? IF you are ignoring the error with optional authmode, you shouldn't get this error.

roxlu commented:
@RonEld I've been looking into this a bit more and currently I think this could be caused because I'm passing wrong buffers into mbedtls. This could be the case but I've been testing/using the dtls_srtp_support PR for the last ~6 weeks (I was using 9ebb178bb7cb1b23ca734c72450f51840acaec9b but I can't find that commit anymore). 

So let's assume it's not an issue related to my code. If that's the case then this paste is interesting:  https://gist.github.com/roxlu/7e870f739876d51060aa673816cdd0de#file-mbedtls-log-L67. There you can see that it detects a bad chipher change message. When you look at the code which flags this error:  https://gist.github.com/roxlu/7e870f739876d51060aa673816cdd0de#file-ssl_tls-c-L15 you see that it's triggered because of a wrong message type. The type that was parsed from by mbedtls is 15, which is a "Certificate Verify".  I'm using   `mbedtls_ssl_conf_authmode(&ssl_conf, MBEDTLS_SSL_VERIFY_OPTIONAL);` which should have disabled this as per your previous comment.

RonEld commented:
@roxlu The commit you referenced has changed for some reason to https://github.com/ARMmbed/mbedtls/pull/1813/commits/15179bfbaa794506c06f923f85d7c71f0dfd89e9
Probably do to rebase I have done or something else.
The only code changing commit since, is my commit for enforcing the certificate messages, which the patch should resolve.
Are you using `mki`? I have just noticed that the mki value is a bit buggy, I will work on it.

RonEld commented:
In addition, the `certificate_verify` message shouldn't have been disabled. The verification *result* itself should have been ignored. The `certificate_verify` message is still mandatory in RFC 5764.

roxlu commented:
@RonEld Ok, when using the commit (which was gone) I did not see a "Certificate Verify" message.  When I'm currently setting auth mode to `..._OPTIONAL`  but also testing with `..._NONE`.  And I would expect this code to be executed then https://github.com/RonEld/mbedtls/blob/dtls_srtp_support/library/ssl_tls.c#L4448 which is not the case. 

I'm not using MKI as far as I'm aware of.

RonEld commented:
@roxlu the code you are refering will not be executec in this case, because `authomde` is *not* `MBEDTLS_SSL_VERIFY_NONE `.
You should verify that both peers work with same version of the SW. For example, have youupdated only one peer of the negotiation? This would explain why you don't expect the `verify_certificate` message.

RonEld commented:
Added an open questoin in the description, adding to comment as well, for discussion:
>How will the server decide whether to accept the MKI value from the client, in the use_srtp extension? Should there be a callback function? Should the server user set the MKI in advance and the server send it if it matches? Should the server just accept the MKI from the client and use it as its MKI ( basically making mbedtls_ssl_dtls_srtp_set_mki_value() a client only function)?

RonEld commented:
I fixed some MKI related issues. Among them, answering my open question, deciding that the server will accept and use the MKI value received from the client.
Comments are welcome

RonEld commented:
Added some interop tests in `ssl-opts.sh` with gnutls and openssl

RonEld commented on include/mbedtls/ssl.h:
```diff
@@ -2291,6 +2355,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \param conf              SSL configuration
+ * \param support_mki_value Enable or disable (MBEDTLS_SSL_DTLS_SRTP_MKI_UNSUPPORTED or
+ *                                    MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED)
```
Fix alignment

RonEld commented on include/mbedtls/ssl.h:
```diff
@@ -2291,6 +2355,73 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *
+ * \return         0 on success, MBEDTLS_ERR_SSL_BAD_INPUT_DATA or MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE
+ */
+int mbedtls_ssl_dtls_srtp_set_mki_value( mbedtls_ssl_context *ssl, unsigned char* mki_value, size_t mki_len );
```
style: change to `unsigned char *mki_value`

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+     *
+     */
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED &&
+                ssl->dtls_srtp_info.mki_len != 0 )
```
fix alignment

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+     * } UseSRTPData;
+
+     * SRTPProtectionProfile SRTPProtectionProfiles<2..2^16-1>;
+     *
```
remove extra line

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    {
+        mki_len = ssl->dtls_srtp_info.mki_len;
+    }
+    /* Extension length = 2bytes for profiles length, ssl->conf->dtls_srtp_profile_list_len*2 (each profile is 2 bytes length ) + 1 byte for srtp_mki vector length and the mki_len value */
```
check that doesn't exceed 80 chars

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    *p++ = (unsigned char)( ( ( 2 + 2*(ssl->conf->dtls_srtp_profile_list_len) + 1 + mki_len ) >> 8 ) & 0xFF );
+    *p++ = (unsigned char)( ( ( 2 + 2*(ssl->conf->dtls_srtp_profile_list_len) + 1 + mki_len )      ) & 0xFF );
+
+
```
remove extra line

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    *p++ = (unsigned char)( ( ( 2*(ssl->conf->dtls_srtp_profile_list_len) ) >> 8 ) & 0xFF );
+    *p++ = (unsigned char)( ( 2*(ssl->conf->dtls_srtp_profile_list_len) ) & 0xFF );
+
+    for( protection_profiles_index=0; protection_profiles_index < ssl->conf->dtls_srtp_profile_list_len; protection_profiles_index++ )
```
fix alignment

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,101 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    }
+
+    /* total extension length: extension type (2 bytes) + extension length (2 bytes) + protection profile length (2 bytes) + 2*nb protection profiles + srtp_mki vector length(1 byte)*/
+    *olen = 2 + 2 + 2 + 2*( ssl->conf->dtls_srtp_profile_list_len ) + 1 + mki_len;
```
add space before and after `*`

RonEld commented on library/ssl_cli.c:
```diff
@@ -1384,6 +1487,115 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     * } UseSRTPData;
+
+     * SRTPProtectionProfile SRTPProtectionProfiles<2..2^16-1>;
+     *
```
remove extra line

RonEld commented on library/ssl_cli.c:
```diff
@@ -1384,6 +1487,115 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     *
+     */
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED &&
+                ssl->dtls_srtp_info.mki_len != 0 )
```
fix alignment

RonEld commented on library/ssl_cli.c:
```diff
@@ -1384,6 +1487,115 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+    /*
+     * get the server protection profile
+     */
+    if (((uint16_t)( ( buf[0]<<8 ) | buf[1] ) ) != 0x0002) { /* protection profile length must be 0x0002 as we must have only one protection profile in server Hello */
```
move comment before the `if` statement

RonEld commented on library/ssl_cli.c:
```diff
@@ -1384,6 +1487,115 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "found srtp profile: %s", profile_info->name ) );
+        }
+
+        if (server_protection == ssl->conf->dtls_srtp_profile_list[i]) {
```
style:
1. `if (` -> `if(`
2. add space before and after brackets

RonEld commented on library/ssl_cli.c:
```diff
@@ -2657,8 +2879,20 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
-        /* Current message is probably the ServerHelloDone */
-        ssl->keep_current_message = 1;
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    /* check if we have a chosen srtp protection profile */
```
align comment with the ext statement

RonEld commented on library/ssl_cli.c:
```diff
@@ -3141,9 +3375,20 @@ static int ssl_write_certificate_verify( mbedtls_ssl_context *ssl )
-        ssl->state++;
-        return( 0 );
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    /* check if we have a chosen srtp protection profile */
```
align comment with the next atatement

RonEld commented on library/ssl_srv.c:
```diff
@@ -648,6 +648,106 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+static int ssl_parse_use_srtp_ext( mbedtls_ssl_context *ssl,
+                               const unsigned char *buf, size_t len )
```
align parameters

RonEld commented on library/ssl_srv.c:
```diff
@@ -648,6 +648,106 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     * } UseSRTPData;
+
+     * SRTPProtectionProfile SRTPProtectionProfiles<2..2^16-1>;
+     *
```
remove extra line

RonEld commented on library/ssl_srv.c:
```diff
@@ -648,6 +648,106 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+    if( ( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED ) &&
+          ( len > ( profile_length + 2 ) ) )
+    {
+        ssl->dtls_srtp_info.mki_len = buf[ profile_length + 2 ];
```
no space next to square brackets

RonEld commented on library/ssl_srv.c:
```diff
@@ -648,6 +648,106 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+            return( MBEDTLS_ERR_SSL_BAD_HS_SERVER_HELLO );
+        }
+
+        ssl->dtls_srtp_info.mki_len = buf[ profile_length + 2 ];
```
no space next to square brackets

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP ) && defined(MBEDTLS_SSL_PROTO_DTLS)
+static void ssl_write_use_srtp_ext( mbedtls_ssl_context *ssl,
+                                unsigned char *buf, size_t *olen )
```
align parameters

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "server hello, adding use_srtp extension" ) );
+
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED &&
+                ssl->dtls_srtp_info.mki_len != 0 )
```
fix alignment

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+    buf[5] = 0x02;
+    switch (ssl->dtls_srtp_info.chosen_dtls_srtp_profile) {
+        case MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80:
+            buf[6] = (unsigned char)( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE >> 8) & 0xFF );
```
style: `8)` -> `8 )`

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+            buf[7] = (unsigned char)( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE     ) & 0xFF );
+            break;
+        case MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32:
+            buf[6] = (unsigned char)( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32_IANA_VALUE >> 8) & 0xFF );
```
style: `8)` -> `8 )`

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+            buf[7] = (unsigned char)( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32_IANA_VALUE     ) & 0xFF );
+            break;
+        case MBEDTLS_SRTP_NULL_HMAC_SHA1_80:
+            buf[6] = (unsigned char)( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_80_IANA_VALUE >> 8) & 0xFF );
```
style: `8)` -> `8 )`

RonEld commented on library/ssl_srv.c:
```diff
@@ -2308,6 +2417,69 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+            buf[7] = (unsigned char)( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_80_IANA_VALUE     ) & 0xFF );
+            break;
+        case MBEDTLS_SRTP_NULL_HMAC_SHA1_32:
+            buf[6] = (unsigned char)( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_32_IANA_VALUE >> 8) & 0xFF );
```
style: `8)` -> `8 )`

RonEld commented on library/ssl_tls.c:
```diff
@@ -644,6 +644,30 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
+         */
+        ssl->dtls_srtp_info.dtls_srtp_keys_len = MBEDTLS_DTLS_SRTP_MAX_KEY_MATERIAL_LENGTH;
+
+        //ssl->dtls_srtp_info.dtls_srtp_keys = (unsigned char *)mbedtls_calloc(1, ssl->dtls_srtp_info.dtls_srtp_keys_len);
```
remove this comment and the next extra line

RonEld commented on library/ssl_tls.c:
```diff
@@ -644,6 +644,30 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
+        //ssl->dtls_srtp_info.dtls_srtp_keys = (unsigned char *)mbedtls_calloc(1, ssl->dtls_srtp_info.dtls_srtp_keys_len);
+
+        ret = handshake->tls_prf( session->master, 48, "EXTRACTOR-dtls_srtp",
+                        handshake->randbytes, 64, ssl->dtls_srtp_info.dtls_srtp_keys, ssl->dtls_srtp_info.dtls_srtp_keys_len );
```
fix alignment

RonEld commented on library/ssl_tls.c:
```diff
@@ -644,6 +644,30 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+    /* check if we have a chosen srtp protection profile */
+    if ( ssl->dtls_srtp_info.chosen_dtls_srtp_profile != MBEDTLS_SRTP_UNSET_PROFILE ) {
+        /* derive key material for srtp session RFC5764 section 4.2 */
```
remove `*/` and `/*` in the next line

RonEld commented on library/ssl_tls.c:
```diff
@@ -379,7 +379,7 @@ static int tls_prf_generic( mbedtls_md_type_t md_type,
     size_t nb;
     size_t i, j, k, md_len;
-    unsigned char tmp[128];
+    unsigned char tmp[144];
```
check why changed to 144?

RonEld replied:
without the change, the function fails because of:
~~~
    if( sizeof( tmp ) < md_len + strlen( label ) + rlen )
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
~~~
`md_len` is 48 for SHA384 (when `MBEDTLS_SHA512_C` defined), `strlen("EXTRACTOR-dtls_srtp")` is 19 anmd `rlen` is 64. This sums to 131.
I'll change to 132, for now

RonEld commented on library/ssl_tls.c:
```diff
@@ -4370,7 +4405,7 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
     size_t i, n;
     const mbedtls_ssl_ciphersuite_t *ciphersuite_info = ssl->transform_negotiate->ciphersuite_info;
-    int authmode = ssl->conf->authmode;
+    int authmode;
```
return assignment, and use `authmode` in checks

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    conf->dtls_srtp_mki_support = support_mki_value;
+}
+
+int mbedtls_ssl_dtls_srtp_set_mki_value( mbedtls_ssl_context *ssl, unsigned char* mki_value, size_t mki_len )
```
style: change `char* mki_value` to `char *mki_value`

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    return 0;
+}
+
+int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf, const mbedtls_ssl_srtp_profile *profiles, size_t profiles_number)
```
align parameters

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    size_t i;
+    /* check in put validity : must be a list of profiles from enumeration */
+    /* maximum length is 4 as only 4 protection profiles are defined */
+    if (profiles_number>4) {
```
style:
1. `if (` -> `if(`
2. add spaces before and after brackets

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    mbedtls_free(conf->dtls_srtp_profile_list);
+    conf->dtls_srtp_profile_list = (mbedtls_ssl_srtp_profile *)mbedtls_calloc(1, profiles_number*sizeof(mbedtls_ssl_srtp_profile));
+
+    for (i=0; i<profiles_number; i++) {
```
style:
1. `for (` -> `for(`
2. add spaces before and after brackets

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+            return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    mbedtls_free(conf->dtls_srtp_profile_list);
```
add spaces next to brackets

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    conf->dtls_srtp_profile_list = (mbedtls_ssl_srtp_profile *)mbedtls_calloc(1, profiles_number*sizeof(mbedtls_ssl_srtp_profile));
+
+    for (i=0; i<profiles_number; i++) {
+        switch (profiles[i]) {
```
style: `switch (profiles[i])` -> `switch( profiles[i] )`

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+                conf->dtls_srtp_profile_list[i] = profiles[i];
+                break;
+            default:
+                mbedtls_free(conf->dtls_srtp_profile_list);
```
add spades next to brackets

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    return( 0 );
+}
+
+mbedtls_ssl_srtp_profile mbedtls_ssl_get_dtls_srtp_protection_profile( const mbedtls_ssl_context *ssl)
```
add space before `)`

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl, unsigned char *key, size_t *key_len ) {
+
+    /* check output buffer size */
+    if ( *key_len < ssl->dtls_srtp_info.dtls_srtp_keys_len) {
```
add space before `)`

RonEld commented on library/ssl_tls.c:
```diff
@@ -6349,6 +6396,102 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+        return MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL;
+    }
+
+    memcpy( key, ssl->dtls_srtp_info.dtls_srtp_keys, ssl->dtls_srtp_info.dtls_srtp_keys_len);
```
add space before `)`

RonEld commented on library/ssl_tls.c:
```diff
@@ -7604,6 +7747,11 @@ void mbedtls_ssl_free( mbedtls_ssl_context *ssl )
 
+#if defined (MBEDTLS_SSL_DTLS_SRTP)
+    mbedtls_platform_zeroize( ssl->dtls_srtp_info.dtls_srtp_keys, ssl->dtls_srtp_info.dtls_srtp_keys_len );
+ //   mbedtls_free( ssl->dtls_srtp_keys );
```
remove

RonEld commented on programs/ssl/ssl_client2.c:
```diff
@@ -344,6 +362,9 @@ struct options
     int etm;                    /* negotiate encrypt then mac?              */
+    int use_srtp;               /* Support SRTP                             */
+    int force_srtp_profile;     /* SRTP protection profile to use or all    */
+    const char* mki;            /* The dtls mki value to use                */
```
change to `char *mki`

RonEld commented on programs/ssl/ssl_client2.c:
```diff
@@ -1334,6 +1377,37 @@ int main( int argc, char *argv[] )
+        if( opt.force_srtp_profile != DFL_SRTP_FORCE_PROFILE )
+        {
+            const mbedtls_ssl_srtp_profile forced_profile[] = { opt.force_srtp_profile };
+            ret = mbedtls_ssl_conf_dtls_srtp_protection_profiles( &conf, forced_profile, sizeof( forced_profile ) / sizeof( mbedtls_ssl_srtp_profile ) );
```
fix alignment

RonEld commented on programs/ssl/ssl_client2.c:
```diff
@@ -1334,6 +1377,37 @@ int main( int argc, char *argv[] )
+                                                                  MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32,
+                                                                  MBEDTLS_SRTP_NULL_HMAC_SHA1_80,
+                                                                  MBEDTLS_SRTP_NULL_HMAC_SHA1_32 };
+            ret = mbedtls_ssl_conf_dtls_srtp_protection_profiles( &conf, default_profiles, sizeof( default_profiles ) / sizeof( mbedtls_ssl_srtp_profile ) );
```
fix alignment

RonEld commented on programs/ssl/ssl_server2.c:
```diff
@@ -2158,9 +2196,45 @@ int main( int argc, char *argv[] )
+        if( opt.force_srtp_profile != DFL_SRTP_FORCE_PROFILE )
+        {
+            const mbedtls_ssl_srtp_profile forced_profile[] = { opt.force_srtp_profile };
+            ret = mbedtls_ssl_conf_dtls_srtp_protection_profiles( &conf, forced_profile, sizeof( forced_profile ) / sizeof( mbedtls_ssl_srtp_profile ) );
```
fix alignment

RonEld commented on programs/ssl/ssl_server2.c:
```diff
@@ -2158,9 +2196,45 @@ int main( int argc, char *argv[] )
+                                                                  MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32,
+                                                                  MBEDTLS_SRTP_NULL_HMAC_SHA1_80,
+                                                                  MBEDTLS_SRTP_NULL_HMAC_SHA1_32 };
+            ret = mbedtls_ssl_conf_dtls_srtp_protection_profiles( &conf, default_profiles, sizeof( default_profiles ) / sizeof( mbedtls_ssl_srtp_profile ) );
```
fix alignment

jeannotlapin commented:
Hi Ron,
any news on merging this PR?
cheers,
Johan

jeannotlapin commented:
@hanno-arm thanks for your review.

I don't have write access to this PR so I hope @RonEld can do the requested modifications on the code(or give me the write access).

About the use of SRTP-DTLS and your comment on the transmission mode not being the one described in the RFC, your understanding on how to proceed by performing the DTLS handshake and then extract the SRTP keys material is correct. I don't think any RTP routing/multiplexing/SSRC association shall be included in the mbedtls library. My guess is that most of the users of the DTLS-SRTP feature already have some RTP stack taking care of this.

I suggest to specify that this PR does not implement the full RFC5764 but the use_srtp DTLS extension described in it allowing the extraction of SRTP key material once the DTLS handshake is completed.

Giving a look on how it is implemented in others libraries which support this extension(not that much actually), both [openssl](https://www.openssl.org/docs/man1.1.1/man3/SSL_set_tlsext_use_srtp.html) and [gnutls](https://www.gnutls.org/manual/html_node/SRTP.html) do the DTLS handshake and key extraction mechanism.

estasig commented on include/mbedtls/ssl.h:
```diff
@@ -2291,6 +2355,85 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+/**
+ * \brief                  Get the generated DTLS-SRTP key material.
+ *                         This function should be called after the handshake is
+ *                         completed. It shall returns 80 bytes of key material
```
Do you mean 60 bytes ? 
```suggestion
 *                         completed. It shall returns 60 bytes of key material
```

RonEld replied:
yes, it's a typo

sbutcher-arm commented:
In discussion with @RonEld, @hanno-arm and @k-stachowiak, this was agreed to be 'approved for design'.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -765,6 +778,42 @@ typedef int mbedtls_ssl_async_resume_t( mbedtls_ssl_context *ssl,
+
+typedef struct mbedtls_dtls_srtp_info_t
+{
+    mbedtls_ssl_srtp_profile chosen_dtls_srtp_profile; /*!< negotiated SRTP profile */
```
If you begin with `/*! ...`, it applies to the next line; if you begin with `/*!< ...`, it applies to the current line.

k-stachowiak replied:
Sorry for nitpicking, but I think it may be useful to remark, that, to be precise, `/*! ...` says that the documented code is _after_ the comment, and `/*!< ...` says it's _before_ the comment. Therefore, the latter type doesn't really need to be on the same line; it may be on the next one. This may come in handy, in case of long comments documenting long lines of code.

hanno-arm replied:
@k-stachowiak Useful nitpicking, thanks!

RonEld commented on include/mbedtls/ssl.h:
```diff
@@ -2291,6 +2355,85 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ */
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl,
+                                            unsigned char *key,
+                                            size_t *key_len );
```
ok, but I really don't like this style

hanno-arm replied:
@RonEld With 3.0 coming up, we have the chance to change this throughout the library, and I think it would be worth brining up this point in an architecture discussion.

hanno-arm replied:
I think it's definitely worth considering, e.g. because it reduces the number of arguments and therefore makes it more likely that compilers can pass them in registers as opposed to stack.

hanno-arm replied:
Ping @mpg

mpg replied:
I think in/out parameters are bad for readability (at least for lengths), so I've actively tried to remove them from our API when moving from 1.3.x to 2.0, as I think readability is extremely important when it comes to buffer sizes. (In practice, their was a case where I initially misread some of our code, then during review Paul misread it in the exact same way, and my conclusion was that it was the in/out length that tripped both of us, hence they were bad.) I've even gone as far as embed this point of view in our [coding standard][1] (search for "limited use of in-out parameters" as we can't link to a section).

[1]: https://tls.mbed.org/kb/development/mbedtls-coding-standards

I'm happy to discuss that again in an architecture discussion for 3.0, but for now I'd rather stick to the current coding standards - as Hanno says, uniformity with the rest of our API matters, if we change it let's do it all at once and after we had a discussion.

(FWIW, some people from the crypto team seem to agree with me, as they are systematically using `uint8_t *, size_t, size_t *` in the signature of functions that write to buffers.)

RonEld replied:
AFAIK, this API in crypto library is used because that's what's used in the tls library, and I have the same argument with the person that agrees with you.

FWIW, when I see a function with a pointer to size as a parameter, I *always* expect it to be initialized with the buffer size. In addition, I am always in favor of reducing number of parameters, as this is much more readable, IMHO, and as @hanno-arm mentioned, compilers can pass them in registers, as opposed to stack.
Of course, for uniformity, I will change the API, but I really think we should consider this for version 3.0

mpg replied:
Thanks for agreeing to change for uniformity for now, and I agree this is a discussion we need to have, as on the readability front we don't have agreement within the team, and even for people who think the 3-argument version is more readable (as I do), there is the point that it might impact stack usage, speed and/or code size, which may or may not outweigh the perceived readability gains.

RonEld commented on include/mbedtls/ssl.h:
```diff
@@ -2291,6 +2355,85 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ * \param key_len          [in/out] key buffer size. outputs the actual number
+ *                         of bytes written
+ *
+ * \return         0 on succes, MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL if the key buffer
```
Actually, as you can see [here](https://github.com/ARMmbed/mbedtls/pull/1813/files#diff-995204bbb15401337db9b64bc314e1f6R6493), this is the only error code returned. But I can change if you wish

RonEld commented on library/ssl_cli.c:
```diff
@@ -2657,8 +2919,19 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
-        /* Current message is probably the ServerHelloDone */
-        ssl->keep_current_message = 1;
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+        /* check if we have a chosen srtp protection profile */
```
[section 4.1](https://tools.ietf.org/html/rfc5764#section-4.1):
~~~
        Client                                           Server

         ClientHello + use_srtp       -------->
                                              ServerHello + use_srtp
                                                        Certificate*
                                                  ServerKeyExchange*
                                                 CertificateRequest*
                                      <--------      ServerHelloDone
         Certificate*
         ClientKeyExchange
         CertificateVerify*
         [ChangeCipherSpec]
         Finished                     -------->
                                                  [ChangeCipherSpec]
                                      <--------             Finished
         SRTP packets                 <------->      SRTP packets
   Note that '*' indicates messages that are not always sent in DTLS.
   The CertificateRequest, client and server Certificates, and
   CertificateVerify will be sent in DTLS-SRTP.
~~~

I'll add this as a comment

RonEld commented on library/ssl_tls.c:
```diff
@@ -379,7 +379,7 @@ static int tls_prf_generic( mbedtls_md_type_t md_type,
     size_t nb;
     size_t i, j, k, md_len;
-    unsigned char tmp[128];
+    unsigned char tmp[132];
```
This change is done to pass the condition (and avoid overflow):
~~~
 if( sizeof( tmp ) < md_len + strlen( label ) + rlen )
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
~~~
In this case, the sizeof `md_len + strlen( label ) + rlen )` is 131, as it's a long label than others.
How do you suggest to name the macro of this tmp buffer?
Perhaps it would be better to assign this buffer on the heap, to avoid future issues as well?

RonEld commented on library/ssl_cli.c:
```diff
@@ -671,6 +671,126 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+     *                         + 2 * number of protection profiles
+     *                         + srtp_mki vector length(1 byte)
+     */
+    *olen = 2 + 2 + 2 + 2 * ( ssl->conf->dtls_srtp_profile_list_len ) + 1 + mki_len;
```
Since the length field is a t the beginning of the message, your suggestion will require either:
1. having a temporary buffer to hold all the data
2.  having `p` point to `buf + 4` and then writing the length to `buf[2]` and `buf[3]`

I am not sure this is a better approach.
I can accept though changing this line to improve readability.( either `p - buf` or `ext_len + 4`)

hanno-arm replied:
@RonEld That's not what I meant. I mean not hardcoding the total length here - you could either add a local variable `total_len` which is incremented as you write the extension, and write that into `*olen` in the end, or alternatively use something like `p - buf` to see how much you have written.
Note that this is different from writing the length values in the target buffer; if you want to do that, you need to - as you say - remember the corresponding part of the buffer until the end. I think that's sensible, too, but requires more work and it not part of the request this comment makes.

RonEld replied:
ok. My initial understanding was to write to the buffer only once everything is validated, but as you said, it's not part of this scope.
I changed it to `p - buf`

RonEld commented on library/ssl_cli.c:
```diff
@@ -1384,6 +1512,130 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+     *  MUST abort the handshake and SHOULD send an invalid_parameter alert.
+     */
+    if( len > 5  &&
+        ( memcmp( ssl->dtls_srtp_info.mki_value, &buf[5], mki_len ) ) )
```
there is a check for [`length == 5 + mki_len`](https://github.com/ARMmbed/mbedtls/pull/1813/files#diff-8d3acce410d251c6ead39efd401db23eR1551)
I only need to verify that `mki_len == buf[4]`

RonEld commented:
@sbutcher-arm I have rebased to resolve conflicts

I have also added a fix to force IPv4 in the GnuTLS client tests in the SRTP, as introduced in 34aa187df6a914d94d56d8b3aeab5692a1a3d59c

jeannotlapin commented:
Hi,
I know this merge is not a priority for mbedtls, but any chance to process it?

Just for information, merging it in release 2.16.1 [see here](https://gitlab.linphone.org/BC/public/external/mbedtls/tree/pr1813-2.16.1) works fine and allows interoperability with webRTC (unless the webRTC end point uses some buggy version of openssl which tends to fragment the DTLS Hello packet which is not supported by mbedtls but this is another issue).

Thanks

RonEld commented:
@jeannotlapin 
Thank you for your information
>I know this merge is not a priority for mbedtls, but any chance to process it?

Yes, this is not one of the highest priority, therefore it keeps getting to the end of our Queue for reviews,unfortunately. As you can see, there are some unresolved conflicts In addition, we have [recently added a feature to export key material](https://github.com/ARMmbed/mbedtls/pull/2538) so this PR will probably need to use the new feature, instead of the way it does it.
You are welcome to continue the necessary  work in this PR, but we can't guarantee on when it will be fully reviewed and approved.

ramatin commented:
Hope it is not forgotten...  Hoping patiently to see some progress... Would be a HUGE benefit for webRTC developers. A few words can keep the hope alive!

RonEld commented on library/ssl_tls.c:
```diff
@@ -763,6 +763,31 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
+         */
+        ssl->dtls_srtp_info.dtls_srtp_keys_len = MBEDTLS_DTLS_SRTP_MAX_KEY_MATERIAL_LENGTH;
+
+        ret = handshake->tls_prf( session->master, 48, "EXTRACTOR-dtls_srtp",
```
This should be removed, in the favor of the new public [mbedtls_ssl_tls_prf](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ssl.h#L4077) function

RonEld commented on library/ssl_tls.c:
```diff
@@ -7590,6 +7644,125 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
+    return( ssl->dtls_srtp_info.chosen_dtls_srtp_profile );
+}
+
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl,
```
This function should be removed. Instead, the application should call `mbedtls_ssl_tls_prf()`

fire commented:
Is this pr still active?

sbutcher-arm commented:
Yes, I think it is. We've been trying to get SRTP into the project for years, and this work has been worked on and off by many people. I don't know if @RonEld is still actively working on it, but if he isn't I don't mind picking it up and championing it in a week or two.

@RonEld - Are you still actively working on this? And if not, would you like me to take it off your hands?

RonEld commented:
@sbutcher-arm I haven't been working on this lately.
Please pick it up.
I would replace the key derivation part to the [TLS PRF key derivation feature](https://tls.mbed.org/kb/how-to/tls_prf)

sbutcher-arm commented:
Thanks for all your work on this @RonEld! I'm happy to take it off you and do the small push to get this over the line.

jeannotlapin commented:
Hi Simon,
any update on this merge? I was planning to add support for SRTP_AEAD_AES_128_GCM and SRTP_AEAD_AES_256_GCM profiles as defined in RFC 7714 section 14.2. I'll do that once the patch is stable and ready to merge (or just after the merge and I'll do a new PR in order not to delay anymore this one).
Thanks.

sbutcher-arm commented:
Hi @jeannotlapin. If you want to add features, I recommend that as separate PRs and given the age of this PR, I personally think it would be better to get this one in first. 

As an update - I haven't started work on this one yet, and plan to soon.

If you wanted to flip this around, and do the rebase yourself and any rework, and I'll review it, you're more than welcome.

jeannotlapin commented:
Thanks @sbutcher-arm 

I'll wait for your rebase and merge before working on the upgrade on a new PR. It's not something I have time to do right now anyway.

jeannotlapin commented:
Hi @sbutcher-arm ,
I just rebased this PR on the current development branch. You can find it on:

https://github.com/jeannotlapin/mbedtls/tree/pr1813

Any chance it would help merging it? Shall I create a new PR from this branch?

cheers

gilles-peskine-arm commented:
Hi @jeannotlapin,

Thanks for staying with us and for following up! Because this is a substantial contribution (and thank you for that), I can't promise that we'll have the bandwidth to review it in the near future. However, if there isn't a pull request with no merge conflicts and passing CI, we definitely won't review it. So since you've done this rebase, please to make a new pull request, and we'll add it to our backlog we're starting to make public at https://github.com/ARMmbed/mbedtls/projects/3.

danh-arm commented:
Superseded by https://github.com/ARMmbed/mbedtls/pull/3235
